// @applitools/eg-socks5-proxy-server/lib/events.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.REMOTE_CONNECTION_TIMEOUT_ERROR = exports.REMOTE_SOCKET_ERROR = exports.ORIGIN_SOCKET_ERROR = exports.REMOTE_PROXY_RFC_1928_REPLY = exports.REMOTE_PROXY_RFC_1928_REPLY_TIMEOUT = exports.REMOTE_PROXY_HANDSHAKE_TIMEOUT = exports.REMOTE_PROXY_HANDSHAKE_COMPLETED = exports.CONNECTED_TO_REMOTE_PROXY = exports.CONNECTED_TO_REMOTE_ADDRESS = exports.ACCEPT_NEW_REQUEST = exports.PROXY_ERROR = exports.PROXY_END = exports.PROXY_DISCONNECT = exports.PROXY_DATA = exports.PROXY_CONNECT = exports.HANDSHAKE = exports.CONNECTION_FILTER = exports.AUTHENTICATION_ERROR = exports.AUTHENTICATION = void 0;
exports.AUTHENTICATION = 'authenticate';
exports.AUTHENTICATION_ERROR = 'authenticateError';
exports.CONNECTION_FILTER = 'connectionFilter';
exports.HANDSHAKE = 'handshake';
exports.PROXY_CONNECT = 'proxyConnect';
exports.PROXY_DATA = 'proxyData';
exports.PROXY_DISCONNECT = 'proxyDisconnect';
exports.PROXY_END = 'proxyEnd';
exports.PROXY_ERROR = 'proxyError';
exports.ACCEPT_NEW_REQUEST = 'accept-new-request';
exports.CONNECTED_TO_REMOTE_ADDRESS = 'connected-to-remote-address';
exports.CONNECTED_TO_REMOTE_PROXY = 'connected-to-remote-proxy';
exports.REMOTE_PROXY_HANDSHAKE_COMPLETED = 'remote-proxy-handshake-completed';
exports.REMOTE_PROXY_HANDSHAKE_TIMEOUT = 'remote-proxy-handshake-timeout';
exports.REMOTE_PROXY_RFC_1928_REPLY_TIMEOUT = 'remote-proxy-rpc-1928-reply-timeout';
exports.REMOTE_PROXY_RFC_1928_REPLY = 'remote-proxy-rpc-1928-reply';
exports.ORIGIN_SOCKET_ERROR = 'orgin-socket-error';
exports.REMOTE_SOCKET_ERROR = 'remote-socket-error';
exports.REMOTE_CONNECTION_TIMEOUT_ERROR = 'remote-connection-timeout-error';
